src/ldapclient_lib.py

```diff
@@ -0,0 +1,381 @@
+        return {
+            "basedn": self.basedn,
+            "ldap_uri": self.ldap_uri,
+            "ldbd_content": self.ldbd_content,
+            "lp_content": self.lp_content,
+        }
+
+    def restore(self, snapshot: dict):
+        """Restore snapshot."""
+        self.basedn = snapshot["basedn"]
+        self.ldap_uri = snapshot["ldap_uri"]
+        self.ldbd_content = snapshot["ldbd_content"]
+        self.lp_content = snapshot["lp_content"]
+
+
+class ConfigDataUnavailableEvent(EventBase):
+    """Charm Event triggered when config data is unavailable."""
+
+    def __init__(
+        self,
+        handle: Handle,
+        api_port: int,
+        ldap_port: int,
+    ):
+        super().__init__(handle)
+        self.api_port = api_port
+        self.ldap_port = ldap_port
+
+    def snapshot(self) -> dict:
+        """Return snapshot."""
+        return {
+            "api_port": self.api_port,
+            "ldap_port": self.ldap_port,
+        }
+
+    def restore(self, snapshot: dict):
+        """Restore snapshot."""
+        self.api_port = snapshot["api_port"]
+        self.ldap_port = snapshot["ldap_port"]
+
+
+class ServerUnavailableEvent(EventBase):
+    """Charm Event triggered When the LDAP server is unavailable."""
+
+
+class LdapReadyEvent(EventBase):
+    """Charm Event triggered when LDAP is ready to start."""
+
+
+class LdapClientProviderCharmEvents(CharmEvents):
+    """Events the LDAP Client requirer can leverage."""
+
+    config_data_unavailable = EventSource(ConfigDataUnavailableEvent)
+    glauth_snap_ready = EventSource(GlauthSnapReadyEvent)
+    server_unavailable = EventSource(ServerUnavailableEvent)
+
+
+class LdapClientRequirerCharmEvents(CharmEvents):
+    """Events the LDAP Client requirer can leverage."""
+
+    certificate_available = EventSource(CertificateAvailableEvent)
+    certificate_unavailable = EventSource(CertificateUnavailableEvent)
+    config_data_available = EventSource(ConfigDataAvailableEvent)
+    config_data_unavailable = EventSource(ConfigDataUnavailableEvent)
+    server_unavailable = EventSource(ServerUnavailableEvent)
+    ldap_ready = EventSource(LdapReadyEvent)
+
+
+class LdapClientProvides(Object):
+    """Provides-side of the ldapclient integration."""
+
+    on = LdapClientProviderCharmEvents()
+
+    def __init__(self, charm: CharmBase, integration_name: str) -> None:
+        super().__init__(charm, integration_name)
+        self.framework.observe(
+            charm.on[integration_name].relation_broken,
+            self._on_relation_broken,
+        )
+        self.framework.observe(
+            charm.on[integration_name].relation_joined,
+            self._on_relation_joined,
+        )
+        self.charm = charm
+        self.integration_name = integration_name
+
+    def _get_hostname(self) -> str:
+        """Get GLAuth hostname."""
+        hostname = subprocess.run(
+            ["cat", "/etc/hostname"], capture_output=True, text=True
```

Comment:
Any particular reason this isn't using `socket.gethostname()` or `platform.node()`?

Reply:
Nope, will swap over to `socket.gethostname()`